content/articles/ml-with-brainjs/index.md

```diff
@@ -0,0 +1,260 @@
+      "text": "it needs more memory",
+      "category": "backend"
+    },
+    {
+      "text": "code with responsive design in users interface",
+      "category": "frontend"
+    },
+    {
+      "text": "navigate the website easily",
+      "category": "frontend"
+    },
+    {
+      "text": "user login and authentication",
+      "category": "backend"
+    },
+    {
+      "text": "forms and dropdowns lists",
+      "category": "frontend"
+    },
+    {
+      "text": "username password email are stored",
+      "category": "backend"
+    },
+    {
+      "text": "programming loading animation",
+      "category": "frontend"
+    },
+    {
+      "text": "mysql, mongo, firebase databases",
+      "category": "backend"
+    },
+    {
+      "text": "restful api is useful with backend",
+      "category": "backend"
+    },
+    {
+        "text": "data access layer is not presentation layer",
+        "category": "backend"
+    },
+    {
+        "text": "the web browser loads dynamic webpages slowly",
+        "category": "frontend"
+    }
+  ]
+```
+
+### Creating the Neural Network
+When building the Neural network, we first need to import the `Brain.js` library into the project. After importing the library, we need to have our data file from where we will access the data to train the model. 
+
+The code snippets below show how we import both the library and the data file.
+
+```js
+/**
+ * bring in the brain.js dependency
+ */
+const brain = require('brain.js')
+/**
+ * Import the data file
+ */
+const data = require('./data.json')
+```
+
+### Long Short Term Memory
+If you have worked with machine learning before, you probably are familiar with the term  Long short-term memory(LSTM). LSTM is an advanced Recurrent Neural Network (RNN) that allows information to stay long before getting lost.  
+
+RNN is a type of neural network that is used for persistent memory. Let's say you are discussing in class, and you remember a concept that the teacher taught in class, and you use the same concepts to solve a problem in your discussion group. The RNN works in the same way  as it remembers the previous information and uses that information to process the current given input. 
+
+The problem with RNN and if that they cannot remember long-term information due to vanishing gradient and for this reason,  LSTM is designed to avoid long-term dependency problems. We are going to use  the inbuilt LSTM in our example to create our network as shown below:
+
+```js
+/**
+ * Create the  neural network
+ */
+const network = new brain.recurrent.LSTM();
+```
+
+### Training the Model
+To train the model, we will take the data from our data file, convert it into an array of values with input and output pairs. We will train our model by running several interactions through the data pair. 
+
+This procedure will take time depending on the number of iteration you perform on the dataset. The more the number of iterations, the higher the accuracy and time it takes.
```

Comment:
This is not correct/accurate. @jamila-laureen 
If the learning rate is small enough, then training a neural network with too many epochs will result in loss converging to one of its local minimum since at every epoch we are moving towards minimum. Once a minimum is reached, then further training will not change the loss since the parameters are updated as Θ←Θ−α∇ΘL and  ∇ΘL=0  at minimum. In other words, the parameters  Θ  does not change.

However, in most cases, we do not want the loss to reach its minimum value since that would result in a neural network which is more prone to over-fitting. The loss function is a measure of how well the neural network does on the training set. If the loss is minimised beyond a certain point, then the neural network is likely to do worse on the unseen examples, i.e., the neural network will have poor generalization.

In other words, as the number of epochs increases, at first both the training error and the test error start decreasing, but after a certain point, the test error starts to increase (the training error continues to decrease since we are minimizing the training error). You can observe this in the following plot of learning curves:

![](https://user-images.githubusercontent.com/63211822/129476202-26f17d02-7c5b-4d41-928b-4e3ef55e36ff.jpg)

The main goal of training a neural network is to do well on unseen examples, not on training examples. That is, we want the neural network to have good generalization.